File: apps/go-api/internal/metrics/metrics.go
// Package metrics is the three-layer telemetry surface:
//
//	api      — HTTP routes
//	backend  — application ops (HTTP or NATS)
//	storage  — adapter calls into the database
//
// Add a new op by using the same names in ObserveBackend / ObserveStorage.
// Native DB exporters are scraped separately by Prometheus.
package metrics

import (
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"

	"prism/go-api/internal/apperr"
)

const Backend = "go"

var (
	APIRequests = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "prism_api_requests_total",
		Help: "HTTP requests by route",
	}, []string{"backend", "storage", "route", "method", "status"})

	APIDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "prism_api_request_duration_seconds",
		Help:    "HTTP request latency",
		Buckets: prometheus.DefBuckets,
	}, []string{"backend", "storage", "route", "method"})

	BackendOps = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "prism_backend_ops_total",
		Help: "Application operations",
	}, []string{"backend", "storage", "op", "source", "result"})

	BackendDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "prism_backend_op_duration_seconds",
		Help:    "Application operation latency",
		Buckets: prometheus.DefBuckets,
	}, []string{"backend", "storage", "op", "source"})

	BackendItems = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "prism_backend_items_total",
		Help: "Points written or samples returned",
	}, []string{"backend", "storage", "op", "source"})

	StorageOps = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "prism_storage_ops_total",
		Help: "Storage adapter operations",
	}, []string{"backend", "storage", "op", "result"})

	StorageDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "prism_storage_op_duration_seconds",
		Help:    "Storage adapter latency",
		Buckets: prometheus.DefBuckets,
	}, []string{"backend", "storage", "op"})

	StorageUp = promauto.NewGaugeVec(prometheus.GaugeOpts{
		Name: "prism_storage_up",
		Help: "1 if the last storage ping succeeded",
	}, []string{"backend", "storage"})
)

func ObserveBackend(storage, op, source string, items int, d time.Duration, err error) {
	BackendOps.WithLabelValues(Backend, storage, op, source, apperr.Result(err)).Inc()
	BackendDuration.WithLabelValues(Backend, storage, op, source).Observe(d.Seconds())
	if err == nil && items > 0 {
		BackendItems.WithLabelValues(Backend, storage, op, source).Add(float64(items))
	}
}

func ObserveStorage(storage, op string, d time.Duration, err error) {
	StorageOps.WithLabelValues(Backend, storage, op, apperr.Result(err)).Inc()
	StorageDuration.WithLabelValues(Backend, storage, op).Observe(d.Seconds())
	if op == "ping" {
		up := 0.0
		if err == nil {
			up = 1
		}
		StorageUp.WithLabelValues(Backend, storage).Set(up)
	}
}

func ObserveAPI(storage, route, method, status string, d time.Duration) {
	APIRequests.WithLabelValues(Backend, storage, route, method, status).Inc()
	APIDuration.WithLabelValues(Backend, storage, route, method).Observe(d.Seconds())
}

File: apps/rust-api/src/metrics.rs
use std::time::Duration;

use metrics::{counter, gauge, histogram};
use metrics_exporter_prometheus::{PrometheusBuilder, PrometheusHandle};

pub const BACKEND: &str = "rust";

pub fn install() -> PrometheusHandle {
    PrometheusBuilder::new()
        .set_buckets(&[
            0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0,
        ])
        .expect("prometheus buckets")
        .install_recorder()
        .expect("prometheus recorder")
}

fn result_label(err: bool) -> &'static str {
    if err {
        "error"
    } else {
        "ok"
    }
}

pub fn observe_api(storage: &str, route: &str, method: &str, status: &str, d: Duration) {
    let labels = [
        ("backend", BACKEND.to_string()),
        ("storage", storage.to_string()),
        ("route", route.to_string()),
        ("method", method.to_string()),
        ("status", status.to_string()),
    ];
    counter!("prism_api_requests_total", &labels).increment(1);
    histogram!(
        "prism_api_request_duration_seconds",
        "backend" => BACKEND.to_string(),
        "storage" => storage.to_string(),
        "route" => route.to_string(),
        "method" => method.to_string()
    )
    .record(d.as_secs_f64());
}

pub fn observe_backend(storage: &str, op: &str, source: &str, items: usize, d: Duration, err: bool) {
    counter!(
        "prism_backend_ops_total",
        "backend" => BACKEND.to_string(),
        "storage" => storage.to_string(),
        "op" => op.to_string(),
        "source" => source.to_string(),
        "result" => result_label(err).to_string()
    )
    .increment(1);
    histogram!(
        "prism_backend_op_duration_seconds",
        "backend" => BACKEND.to_string(),
        "storage" => storage.to_string(),
        "op" => op.to_string(),
        "source" => source.to_string()
    )
    .record(d.as_secs_f64());
    if !err && items > 0 {
        counter!(
            "prism_backend_items_total",
            "backend" => BACKEND.to_string(),
            "storage" => storage.to_string(),
            "op" => op.to_string(),
            "source" => source.to_string()
        )
        .increment(items as u64);
    }
}

pub fn observe_storage(storage: &str, op: &str, d: Duration, err: bool) {
    counter!(
        "prism_storage_ops_total",
        "backend" => BACKEND.to_string(),
        "storage" => storage.to_string(),
        "op" => op.to_string(),
        "result" => result_label(err).to_string()
    )
    .increment(1);
    histogram!(
        "prism_storage_op_duration_seconds",
        "backend" => BACKEND.to_string(),
        "storage" => storage.to_string(),
        "op" => op.to_string()
    )
    .record(d.as_secs_f64());
    if op == "ping" {
        gauge!(
            "prism_storage_up",
            "backend" => BACKEND.to_string(),
            "storage" => storage.to_string()
        )
        .set(if err { 0.0 } else { 1.0 });
    }
}
